File: Homework1/python/LCS.py
'''
@Description: 
@Author: Zhaoxi Chen
@Github: https://github.com/FrozenBurning
@Date: 2020-03-13 23:29:13
@LastEditors: Zhaoxi Chen
@LastEditTime: 2020-03-14 10:41:09
'''
import sys
from argparse import ArgumentParser

def score_criterion(char1, char2,alpha,beta,gamma):
    '''
    @description: score computing between 2 matrix
    @param 
    @return: score
    '''
    if char1 == '-' or char2 == '-': #gap
        return gamma
    elif char1 != char2:   #mismatch
        return beta
    elif char1 == char2:   # match
        return alpha

def scoring_with_matrix(seq1, seq2,alpha,beta,gamma):
    '''
    @description: NW algorithms
    @param  
    @return: score matrix & trace matrix 
    '''
    seq1 = '-' + seq1
    seq2 = '-' + seq2
    score_mat = {}
    trace_mat = {}

    for i,char1 in enumerate(seq1):
        score_mat[i] = {}
        trace_mat[i] = {}
        for j,char2 in enumerate(seq2):
            if i == 0:                    # first row, gap in seq1
                score_mat[i][j] = j*gamma
                trace_mat[i][j] = [1]
                continue
            if j == 0:                    # first column, gap in seq2
                score_mat[i][j] = i*gamma
                trace_mat[i][j] = [2]
                continue
            ul = score_mat[i-1][j-1] + score_criterion(char1, char2,alpha,beta,gamma)     # from up-left, mark 0
            l  = score_mat[i][j-1]   + score_criterion('-', char2,alpha,beta,gamma)   # from left, mark 1, gap in seq1
            u  = score_mat[i-1][j]   + score_criterion(char1, '-',alpha,beta,gamma)   # from up, mark 2, gap in seq2
            chosen = max([ul,l,u])
            score_mat[i][j] = chosen
            trace_mat[i][j] = [ind for ind,s in enumerate([ul,l,u]) if s==chosen]
            # trace_mat[i][j] = [ul, l, u].index(chosen)   # record which direction
    return score_mat, trace_mat



pathcode=[]
def traceback(seq1, seq2, trace_mat,path_code=''):
    '''
    @description: traceback optimal path
    @param trace_mat: generated from scoring_with_matrix
    @return: path_code
    '''
    seq1, seq2 = '-' + seq1, '-' + seq2
    i, j = len(seq1) - 1, len(seq2) - 1
    # path_code = ''
    if i>0 or j>0:
        direction = trace_mat[i][j] #supposed to be list
        for ind,direct in enumerate(direction):
            tmp_i=i
            tmp_j=j
            if direct == 0:                    # from up-left direction
                tmp_i = i-1
                tmp_j = j-1
                tmppath_code = '0' + path_code
            elif direct == 1:                  # from left
                tmp_j = j-1
                tmppath_code = '1' + path_code
            elif direct == 2:                  # from up
                tmp_i = i-1
                tmppath_code = '2' + path_code

            if tmp_i >0 or tmp_j>0:
                traceback(seq1[1:tmp_i+1],seq2[1:tmp_j+1],trace_mat,tmppath_code)
            else:
                pathcode.append(tmppath_code)            
    else:
        # pathcode.append(path_code)
        return 

def print_m(seq1, seq2, m):
    """print score matrix or trace matrix"""
    seq1 = '-' + seq1; seq2 = '-' + seq2
    print()
    print(' '.join(['%3s' % i for i in ' '+seq2]))
    for i, p in enumerate(seq1):
        line = [p] + [m[i][j] for j in range(len(seq2))]
        print(' '.join(['%3s' % i for i in line]))
    print()
    return

def pretty_print_align(seq1, seq2, path_code):
    '''
    return pair alignment result string from
    path code: 0 for match, 1 for gap in seq1, 2 for gap in seq2
    '''
    align1 = ''
    middle = ''
    align2 = ''
    for p in path_code:
        if p == '0':
            align1 = align1 + seq1[0]
            align2 = align2 + seq2[0]
            if seq1[0] == seq2[0]:
                middle = middle + '|'
            else:
                middle = middle + ' '
            seq1 = seq1[1:]
            seq2 = seq2[1:]
        elif p == '1':
            align1 = align1 + '-'
            align2 = align2 + seq2[0]
            middle = middle + ' '
            seq2 = seq2[1:]
        elif p == '2':
            align1 = align1 + seq1[0]
            align2 = align2 + '-'
            middle = middle + ' '
            seq1 = seq1[1:]

    print('Alignment:\n\n   ' + align1 + '\n   ' + middle + '\n   ' + align2 + '\n')
    return


def main():
    parser = ArgumentParser(description='Needleman Wunsch Algorithm')
    # parser.add_argument('--seq1',type=str,required=True,help='input sequence 1')
    # parser.add_argument('--seq2',type=str,required=True,help='input sequence 2')
    parser.add_argument('--seq1',type=str,required=False,default='AAGC',help='input sequence 1')
    parser.add_argument('--seq2',type=str,required=False,default='AGT',help='input sequence 2')
    parser.add_argument('-a','--alpha',type=int,required=False,default=1)
    parser.add_argument('-b','--beta',type=int,required=False,default=-1)
    parser.add_argument('-g','--gamma',type=int,required=False,default=-1)
    parser.add_argument('-f','--file',type=str,required=False,default='',help='input test file including 2 strings')
    args = parser.parse_args()
    if not len(args.file) == 0: 
        with open(args.file, 'r') as f:
            strings = f.read().splitlines()
        seq1 = strings[0]
        seq2 = strings[1]
    else:
        try:
            seq1,seq2 = args.seq1,args.seq2
        except:
            print("Invalid Inputs")
    print('current params: alpha=%.1f,beta=%.1f,gamma=%.1f' % (args.alpha, args.beta,  args.gamma))
    print('sequence 1: %s' % seq1)
    print('sequence 2: %s' % seq2)
    # print('args:',args.gamma)
    score_mat, trace_mat = scoring_with_matrix(seq1, seq2,args.alpha,args.beta,args.gamma)
    print_m(seq1, seq2, score_mat)
    print_m(seq1, seq2, trace_mat)

    traceback(seq1, seq2, trace_mat)
    
    for i in range(0,len(pathcode)):
        pretty_print_align(seq1, seq2, pathcode[i])
    print(pathcode)
    # print('   '+path_code)

    print('total score:',score_mat[len(seq1)][len(seq2)])

if __name__ == '__main__':
    main()

File: Homework1/cpp/LCS.cpp
/*
 * @Description: 
 * @Author: Zhaoxi Chen
 * @Github: https://github.com/FrozenBurning
 * @Date: 2020-03-13 23:31:23
 * @LastEditors: Zhaoxi Chen
 * @LastEditTime: 2020-03-14 10:46:25
 */
#include <iostream>
#include <string>
#include <fstream>

using namespace std;

int NeedlemanWunsch(string seq1, string seq2, int alpha, int beta, int gamma)
{
	char gap = 0;
	// Read data
	int len1 = seq1.size();
	int len2 = seq2.size();
    // trans string into vector
	char * seq1_vec = new char [len1+1];
	char * seq2_vec = new char [len2+1];
	for (int i = 0; i < len1; i++)
    {
        seq1_vec[i] = seq1[i];
    }
	for (int i = 0; i < len2; i++)
    {
        seq2_vec[i] = seq2[i];
    }

	// Initialize score matrix
	int ** score_mat = new int * [len1+1];

	// Initialize multiple auxiliary matrix
	int ** tracex = new int * [len1+1];
	int ** tracey = new int * [len1+1];
    int ** recorder1 = new int * [len1+1];
	int ** recorder2 = new int * [len1+1];

	for (int i = 0; i < len1+1; i++)
    {
		score_mat[i] = new int [len2+1];
		score_mat[i][0] = i*gamma;
		tracex[i] = new int [len2+1];
		tracey[i] = new int [len2+1];   
        recorder1[i] = new int [len2+1];	
		recorder2[i] = new int [len2+1];	
    }

	for (int i = 0; i < len2+1; i++) 
	{
		score_mat[0][i] = i*gamma;
	}

	for (int i = 1; i < len1+1; i++)
	{
		tracex[i][0] = i-1;
		tracey[i][0] = 0;
		recorder1[i][0] = seq1_vec[i-1];
		recorder2[i][0] = gap;
	}
	for (int i = 1; i < len2+1; i++)
	{
		tracex[0][i] = 0;
		tracey[0][i] = i-1;
		recorder1[0][i] = gap;
		recorder2[0][i] = seq2_vec[i-1];
	}
	tracex[0][0] = -1;
	tracey[0][0] = -1;



    // Main Solver
	for (int i = 1; i < len1+1; i++)
	{
		for (int j =1; j < len2+1; j++)
		{
			int diag, left, up,candidate;
			// compute scores in three directions
			if (seq1_vec[i-1] == seq2_vec[j-1]) 
			{
				diag = score_mat[i-1][j-1] + alpha;
			}
			else
			{
				diag = score_mat[i-1][j-1] + beta;
			}
			left = score_mat[i-1][j] + gamma;
			up = score_mat[i][j-1] + gamma;

            // update decision 
            candidate = max(diag,max(left,up));
            if (candidate == left)
            {
                score_mat[i][j] = left;
				tracex[i][j] = i-1;
				tracey[i][j] = j;
				recorder1[i][j] = seq1_vec[i-1];
				recorder2[i][j] = gap;
            }
            else if (candidate==up)
            {
				score_mat[i][j] = up;
				tracex[i][j] = i;
				tracey[i][j] = j-1;
				recorder1[i][j] = gap;
				recorder2[i][j] = seq2_vec[j-1];                
            }
            else
            {
				score_mat[i][j] = diag;
				tracex[i][j] = i-1;
				tracey[i][j] = j-1;
				recorder1[i][j] = seq1_vec[i-1];
				recorder2[i][j] = seq2_vec[j-1];
            }
		}
	}

	int row_ind = len1;
	int col_ind = len2;
	int * backtrace_row = new int [len1+len2];
	int * backtrace_col = new int [len1+len2];

	int counter = 0;
    int tmpx,tmpy;
	backtrace_row[counter] = row_ind;
	backtrace_col[counter] = col_ind;
	while(row_ind != 0 || col_ind != 0)
	{
		counter ++;
		tmpx = tracex[row_ind][col_ind];
		tmpy = tracey[row_ind][col_ind];
		row_ind = tmpx;
		col_ind = tmpy;
		backtrace_row[counter] = row_ind;
		backtrace_col[counter] = col_ind;
	}


	FILE * f = fopen("../Answer.txt", "w");
	for (int i = counter-1; i >= 0; i--)
	{
		char output1 = recorder1[backtrace_row[i]][backtrace_col[i]];
		char output2 = recorder2[backtrace_row[i]][backtrace_col[i]];
		int same = output1*output2;
		if (same != 0)
			fprintf(f,"%c",output1);
	}
	fclose(f);

	return(score_mat[len1][len2]);
}

int main()
{
    string seq1,seq2;
    // parameters
	int alpha = 1;
	int beta = 0;
	int gamma = 0;

	ifstream LCSfile("../LongestCommonSeq.txt");
	getline(LCSfile, seq1);
	getline(LCSfile, seq2);
	LCSfile.close();

	int score = NeedlemanWunsch(seq1,seq2, alpha, beta, gamma);
	printf("Score: %d\n",score);

	return 0;
}